### ISSUE TITLE
Should array classes be created at runtime?

### ISSUE THREAD

**Issue by volth** [konsoletyper/teavm#412]
It seems that it is not possible to get full lists of array types at compile time, because of arrays can be created using reflection (and often are).

Testcase:
```scala
class X(x: Int)
val a = Array.tabulate[X](20)(new X(_))
```

Using of `Array[X]` is not detected at compile time, so `X`'s `.arrayType` is NULL which result in crash at
https://github.com/konsoletyper/teavm/blob/90e00f7eb4c2333dc69eb05a64daa4fd9d46a95a/classlib/src/main/java/org/teavm/classlib/java/lang/reflect/TArray.java#L72

**Comment by konsoletyper** [comment]
No, they should not be (and can't be) created at runtime. Instead, results of static analysis should be propagated to C code generator and all necessary array should be precompiled. For now reflection support in TeaVM/C is much more limited than in TeaVM/JS

**Comment by konsoletyper** [comment]
As a workaround you can instantiate empty array of X somewhere in `main` method.

**Comment by volth** [comment]
I assumed that the structures in `$outdir/arrays` are boilerplate and a small runtime function could make them on demand, so any class will always have `.arrayType`

**Comment by konsoletyper** [comment]
No, there are other reasons why it's impossible. Class is encoded in object as a 29-bit integer, and this only works if all classes are no more than 2^32 bytes between each other. Currently, all classes are in the same data segment, that makes this requirement met. If we construct new classes somewhere in heap, we can't guarantee that OS allocates heap in an address that is at least 2^32 bytes close to the area where is loaded data segment of executable file. This can be solved by copying all classes to the heap during initialization. However, this should slow down initialization and increase memory consumption.

On the other hand arrays of depth more than 3 are rarely used. Also, coding pattern that can make infinite upper bound during static analysis is rare. So we can avoid latter case: assume that all arrays of depth more than 3 are `Object[][][]`. I think with this all assumptions we can only break things in quite rare practical cases.
